saleor/graphql/order/mutations/fulfillments.py

```diff
@@ -308,3 +312,245 @@ def perform_mutation(cls, _root, info, **data):
+                            "exclusive and attempts to call mutation with both options "
+                            "is incorrect."
+                        ),
+                        code=OrderErrorCode.INVALID.value,
+                    ),
+                    "include_shipping_costs": ValidationError(
+                        (
+                            "The amountToRefund and includeShippingCosts are mutually "
+                            "exclusive and attempts to call mutation with both options "
+                            "is incorrect."
+                        ),
+                        code=OrderErrorCode.INVALID.value,
+                    ),
+                }
+            )
+        qs = order_models.Order.objects.prefetch_related("payments")
+        order = cls.get_node_or_error(
+            info, order_id, field="order", only_type=Order, qs=qs
+        )
+        payment = order.get_last_payment()
+        if not payment or not payment.can_refund():
+            raise ValidationError(
+                {
+                    "order": ValidationError(
+                        "Order cannot be refunded.",
+                        code=OrderErrorCode.CANNOT_REFUND.value,
+                    )
+                }
+            )
+
+        if amount_to_refund is not None and amount_to_refund > payment.captured_amount:
+            raise ValidationError(
+                {
+                    "amount_to_refund": ValidationError(
+                        (
+                            "The amountToRefund is greater than the maximal possible "
+                            "amount to refund."
+                        ),
+                        code=OrderErrorCode.CANNOT_REFUND.value,
+                    ),
+                }
+            )
+        cleaned_input = {
+            "amount_to_refund": amount_to_refund,
+            "include_shipping_costs": include_shipping_costs,
+            "order": order,
+            "payment": payment,
+        }
+        order_lines_data = input.get("order_lines")
+        fulfillment_lines_data = input.get("fulfillment_lines")
+
+        if not order_lines_data and not fulfillment_lines_data:
+            raise ValidationError(
+                {
+                    "order_lines": ValidationError(
+                        "The orderLines or fulfillmentLines cannot be empty.",
+                        code=OrderErrorCode.REQUIRED.value,
+                    ),
+                    "fulfillment_lines": ValidationError(
+                        "The orderLines or fulfillmentLines cannot be empty.",
+                        code=OrderErrorCode.REQUIRED.value,
+                    ),
+                }
+            )
+        if order_lines_data:
+            cls.clean_lines(order_lines_data, cleaned_input)
+        if fulfillment_lines_data:
+            cls.clean_fulfillment_lines(fulfillment_lines_data, cleaned_input)
```

Comment:
First you  are checking `if not order_lined_data`, next you are checking `if order data`. You can move check `if not order_line_data` into clean_lines and then raise validation error related to `order_lines` or `fulfillment_lines`instead of rising both

Reply:
This approach would require to pass `fulfillment_lines` to `clean_lines` and order_lines to `clean_fulfillment_lines`. The purpose of this error is raised when both lists are empty.